### DISCUSSION TITLE
Not working inside a function defined globally. Is it possible?

### DISCUSSION THREAD

**Discussion by abnersajr** [Spawnova/ShinsImageScanClass#11] [Q&A] [answered]
I'm new to AHK, not sure about the differences of Function and Subroutine.
This is a pseudo-code but is hos is structured and how I would like to use it.
If I change the functions to subroutines it works as expected.
Not sure if is a bug, or something I'm missing from how AHK works.

Also if I define the class everytime inside the function it works.

Basically I have a code:
```ahk
...
Global scan := new ShinsImageScanClass("Prog") ; works once
;scan := new ShinsImageScanClass("Prog") : never work

x:: {
  ProgLog()
}

hasImage() {
 return has_image := scan.Image("image.png", 0)
}

ProgLog() {
 if (has_image) {
  ; do something
 }
}
...
``` 

Didn't opened as an issue, because maybe is something I don't know properly.
Anyway thx for your job

**Comment by Spawnova** [comment, marked as answer]
Hey there, this class is designed to be initialized just once, creating a new instance every time inside a function would be very poor performance.

The problem here is that the super global name "scan" is actually used inside my class, however I've released a new update that forces local variables inside key functions which should fix that problem, and allow you to define scan globally, you will just have to update to the newest version

Also in ahk you don't use brackets for the hotkeys =p


```ahk
Global scan := new ShinsImageScanClass("Prog")

x::
if (scan.Image("image.png",0,x,y)) {
	msgbox % "Found image at " x "," y
} else {
	msgbox % "Could not find image"
}
return

y::
func()
return


func() { ;can be accessed globally 
	if (scan.Image("image.png",0,x,y)) {
		msgbox % "Found image at " x "," y
		return 1
	} else {
		msgbox % "Could not find image"
		return 0
	}
}
```

hopefully that helps
